process_pr.py

```diff
@@ -1459,7 +1460,50 @@ def process_pr(repo_config, gh, repo, issue, dryRun, cmsbuild_user=None, force=F
             return
 
         print("Processing commits")
+        commit_cache = {k: v for k, v in bot_cache.items() if REGEX_COMMIT_SHA.match(k)}
         if pr.commits < MAX_INITIAL_COMMITS_IN_PR or ok_too_many_commits:
+            seen_commits_cnt = len(commit_cache)
+            if all_commits.totalCount < seen_commits_cnt:
+                print(
+                    "Number of commits in cache ({0}) is greater than the number of commits in PR ({1}) - possible squash detected".format(
+                        seen_commits_cnt, all_commits.totalCount
+                    )
+                )
+
+                last_seen_commit_sha = sorted(
+                    [(k, commit_cache[k]["time"]) for k in commit_cache], key=lambda q: q[1]
+                )[-1][0]
+                new_head_commit_sha = pr.head.sha
+                base_commit_sha = pr.base.sha
+
+                changes_before_squash = repo.compare(base_commit_sha, last_seen_commit_sha)
+                changes_after_squash = repo.compare(base_commit_sha, new_head_commit_sha)
+
+                diff_before_squash = {
+                    (file.filename, file.patch) for file in changes_before_squash.files
+                }
+                diff_after_squash = {
+                    (file.filename, file.patch) for file in changes_after_squash.files
+                }
+
+                if diff_before_squash ^ diff_after_squash:
+                    print("PR diff changed, will not preserve signatures")
+                else:
+                    print("PR diff not changed, preserving signatures")
+                    bot_cache[new_head_commit_sha] = {
+                        "time": bot_cache[last_seen_commit_sha]["time"] + 1,
+                        "files": sorted(
+                            x["filename"]
+                            for x in get_commit(repo.full_name, new_head_commit_sha)["files"]
```

Comment:
@iarspider , this does not look right e.g. lets say we have

- commit1 at T0: touching core and reco files
- core signs PR at T1
- commit2 at T2: touching reco files
- bot see commit2 and it is now our last_seen_commit
- squash_commit at T2+1 will have all files core and reco which means core signatures will not be preserved ... right?